### PR TITLE
feat(expresive-modal): introduce Web Components

### PR DESCRIPTION
**Pull request by asudoh**
### Related Ticket(s)

Refs #2988.

### Description

Introduces `<dds-modal>`, an expressive (DDS) version of modal.
Also introduces `<dds-btn>`, an expressive (DDS) version of button, as it's needed for demonstration.

We could use `<bx-btn>` with applying expressive tokens in application (Storybook) code, but having a tag (custom element) that works as an expressive version out-of-the-box will provide the full benefit of Web Components, which is, "It just works by putting them as DOM elements".

`<dds-btn>` utilizes the expressive token proxy code to minimize the Sass code. Did the following to do that:

* Introduce Sass maps for expressive size/type/component tokens.
* Extracted out and streamlined the code to generate CSS custom properties from expressive size/type/component tokens.
* A fix to the expressive token proxy code for button so it works when it's used within a CSS ruleset (e.g. one for `:host(#{$dds-prefix}-btn)`).

### Changelog

**New**

- `<dds-modal>`, an expressive (DDS) version of modal.
- `<dds-btn>`, an expressive (DDS) version of button, as it's needed for demonstration.
- Sass maps for expressive size/type/component tokens.

**Changed**

- Extracted out and streamlined the code to generate CSS custom properties from expressive size/type/component tokens.
- A fix to the expressive token proxy code for button so it works when it's used within a CSS ruleset (e.g. one for `:host(#{$dds-prefix}-btn)`).

### CODE REVIEW

File: packages/web-components/src/components/modal/modal.scss [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+      .#{$prefix}--modal-container {
+        width: calc(100% - #{$carbon--spacing-07});
+        max-width: calc(100% - #{$carbon--spacing-07});
+        height: calc(100% - #{$carbon--spacing-07});
+        max-height: none;
+      }
+    }
+  }
+
+  .#{$dds-prefix}-ce--modal__hedaer--with-body {
```
Comment (jeffchew) [inline packages/web-components/src/components/modal/modal.scss:51]: @asudoh is this a typo in the classname in the custom elements side?

File: packages/web-components/src/components/modal/modal.ts [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    });
+    const bodyClasses = classMap({
+      [`${ddsPrefix}-ce--modal__body--with-footer`]: hasBody && hasFooter,
+    });
+    return html`
+      <a id="start-sentinel" class="${prefix}--visually-hidden" href="javascript:void 0" role="navigation"></a>
+      <div
+        role="dialog"
+        class="${containerClasses}"
+        tabidnex="-1"
```
Comment (jeffchew) [inline packages/web-components/src/components/modal/modal.ts:154]: `tabindex` mispelling here

File: packages/web-components/src/components/modal/__stories__/README.stories.mdx [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import '@carbon/ibmdotcom-web-components/es/components/modal/modal-heading';
+import '@carbon/ibmdotcom-web-components/es/components/modal/modal-close-button';
+import '@carbon/ibmdotcom-web-components/es/components/modal/modal-body';
+import '@carbon/ibmdotcom-web-components/es/components/modal/modal-footer';
+```
+
+##### HTML
+
+```html
+<dds-modal open>
```
Comment (jeffchew) [inline packages/web-components/src/components/modal/__stories__/README.stories.mdx:29]: @asudoh are there instructions on how to trigger the modal to open and close?

File: packages/web-components/src/components/modal/modal.ts (hunk 2 above)
Comment (asudoh) [reply packages/web-components/src/components/modal/modal.ts]: Good catch thanks! Updated.

File: packages/web-components/src/components/modal/__stories__/README.stories.mdx (hunk 3 above)
Comment (asudoh) [reply packages/web-components/src/components/modal/__stories__/README.stories.mdx:29]: Good question - Added the doc.

File: packages/web-components/src/components/modal/modal.scss (hunk 1 above)
Comment (asudoh) [reply packages/web-components/src/components/modal/modal.scss:51]: `#{$dds-prefix}-ce` prefix is for telling that such CSS class happens only with custom elements implementation, but I'm open for removing such `-ce` prefix.

Comment (jeffchew) [reply packages/web-components/src/components/modal/modal.scss:51]: @asudoh I meant `header` instead of `hedaer`

**Pull request merged**